### PR TITLE
HDFS-15568. namenode start failed to start when dfs.namenode.max.snapshot.limit set.

### PR DESCRIPTION
**Pull request by bshashikant**
please see https://issues.apache.org/jira/browse/HDFS-15568

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  40m 35s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  32m  7s |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 16s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  compile  |   1m  8s |  trunk passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  checkstyle  |   0m 47s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 15s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  17m 51s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 50s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javadoc  |   1m 19s |  trunk passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +0 :ok: |  spotbugs  |   3m 13s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   3m 11s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 11s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javac  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  1s |  the patch passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  javac  |   1m  1s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 40s |  hadoop-hdfs-project/hadoop-hdfs: The patch generated 3 new + 89 unchanged - 0 fixed = 92 total (was 89)  |
| +1 :green_heart: |  mvnsite  |   1m 10s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 27s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 46s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javadoc  |   1m 16s |  the patch passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  findbugs  |   3m 27s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  |  80m 10s |  hadoop-hdfs in the patch passed.  |
| -1 :x: |  asflicense  |   0m 35s |  The patch generated 4 ASF License warnings.  |
|  |   | 209m 18s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.fs.viewfs.TestViewFSOverloadSchemeWithMountTableConfigInHDFS |
|   | hadoop.fs.viewfs.TestViewFileSystemHdfs |
|   | hadoop.fs.viewfs.TestViewFileSystemLinkMergeSlash |
|   | hadoop.hdfs.TestSnapshotCommands |
|   | hadoop.fs.viewfs.TestViewFsAtHdfsRoot |
|   | hadoop.hdfs.TestFileChecksumCompositeCrc |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2296 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 94a92cfd8c29 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / e5fe3262702 |
| Default Java | Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01 |
| checkstyle | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/1/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/1/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/1/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 3060 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirectory.java [hunk 1]
```diff
@@ -508,6 +508,10 @@ FSNamesystem getFSNamesystem() {
     return namesystem;
   }
 
+  public boolean isImageLoaded() {
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirectory.java:511]: Add javadoc

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java [hunk 2]
```diff
@@ -368,6 +368,13 @@ void assertFirstSnapshot(INodeDirectory dir,
     }
   }
 
+  boolean captureOpenFiles() {
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java:371]: javadoc

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java [hunk 3]
```diff
@@ -368,6 +368,13 @@ void assertFirstSnapshot(INodeDirectory dir,
...
+    return captureOpenFiles;
+  }
+
+  int getMaxSnapshotLimit() {
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java:375]: VisibleForTesting

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java [hunk 4]
```diff
@@ -133,4 +137,68 @@ public void testValidateSnapshotIDWidth() throws Exception {
...
+    cluster.waitActive();
+    DistributedFileSystem hdfs = cluster.getFileSystem();
+    hdfs.mkdirs(snapshottableDir);
+    hdfs.allowSnapshot(snapshottableDir);
+    int i = 0;
+    for (; i < numSnapshots; i++) {
+      hdfs.createSnapshot(snapshottableDir, "s" + i);
+    }
+    try {
+      hdfs.createSnapshot(snapshottableDir, "s" + i);
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java:161]: Use LambdaTestUtils#intercept

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java [hunk 5]
```diff
@@ -133,4 +137,68 @@ public void testValidateSnapshotIDWidth() throws Exception {
...
+              "max snapshot limit"));
+    }
+
+    // now change max snapshot directory limit to 2 and restart namenode
+    cluster.getNameNode().getConf().setInt(DFSConfigKeys.
+        DFS_NAMENODE_SNAPSHOT_MAX_LIMIT, 2);
+    cluster.restartNameNodes();
+
+    // make sure edits of all previous 5 create snapshots are replayed
+    Assert.assertEquals(numSnapshots, cluster.getNamesystem().
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java:175]: Extract the getSnapshotManager()

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java [hunk 6]
```diff
@@ -133,4 +137,68 @@ public void testValidateSnapshotIDWidth() throws Exception {
...
+    // make sure namenode has the new snapshot limit configured as 2
+    Assert.assertEquals(2,
+        cluster.getNamesystem().getSnapshotManager().getMaxSnapshotLimit());
+
+    // Any new snapshot creation should still fail
+    try {
+      hdfs.createSnapshot(snapshottableDir, "s" + i);
+      Assert.fail("Expected SnapshotException not thrown");
+    } catch (SnapshotException se) {
+      Assert.assertTrue(
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java:187]: LambdaTestUtils

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java [hunk 7]
```diff
@@ -448,22 +455,31 @@ public String createSnapshot(final LeaseManager leaseManager,
...
+    checkSnapshotLimit(maxSnapshotFSLimit, n);
+    srcRoot.addSnapshot(this, snapshotName, leaseManager, mtime);
       
     //create success, update id
     snapshotCounter++;
     numSnapshots.getAndIncrement();
     return Snapshot.getSnapshotPath(snapshotRoot, snapshotName);
   }
+
+  void checkSnapshotLimit(int limit, int numSnapshots)
```
Comment (szetszwo) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java:467]: I suggest to add limit type to the error message as below.

```
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/DirectorySnapshottableFeature.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/DirectorySnapshottableFeature.java
index 266c0a71241..7a47ab4000d 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/DirectorySnapshottableFeature.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/DirectorySnapshottableFeature.java
@@ -190,8 +190,7 @@ public Snapshot addSnapshot(INodeDirectory snapshotRoot,
           + n + " snapshot(s) and the snapshot quota is "
           + snapshotQuota);
     }
-    snapshotManager.checkSnapshotLimit(snapshotManager.
-        getMaxSnapshotLimit(), n);
+    snapshotManager.checkPerDirectorySnapshotLimit(n);
     final Snapshot s = new Snapshot(id, name, snapshotRoot);
     final byte[] nameBytes = s.getRoot().getLocalNameBytes();
     final int i = searchSnapshot(nameBytes);
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java
index 0a2e18c3dc3..7c482074486 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java
@@ -455,7 +455,7 @@ public String createSnapshot(final LeaseManager leaseManager,
           "snapshot IDs and ID rollover is not supported.");
     }
     int n = numSnapshots.get();
-    checkSnapshotLimit(maxSnapshotFSLimit, n);
+    checkFileSystemSnapshotLimit(n);
     srcRoot.addSnapshot(this, snapshotName, leaseManager, mtime);
       
     //create success, update id
@@ -464,12 +464,19 @@ public String createSnapshot(final LeaseManager leaseManager,
     return Snapshot.getSnapshotPath(snapshotRoot, snapshotName);
   }
 
-  void checkSnapshotLimit(int limit, int numSnapshots)
-      throws SnapshotException {
+  void checkFileSystemSnapshotLimit(int n) throws SnapshotException {
+    checkSnapshotLimit(maxSnapshotFSLimit, n, "file system");
+  }
+
+  void checkPerDirectorySnapshotLimit(int n) throws SnapshotException {
+    checkSnapshotLimit(maxSnapshotLimit, n, "per directory");
+  }
+
+  private void checkSnapshotLimit(int limit, int numSnapshots,
+      String type) throws SnapshotException {
     if (numSnapshots >= limit) {
-      String msg = "there are already " + (numSnapshots + 1)
-          + " snapshot(s) and the max snapshot limit is "
-          + limit;
+      String msg = "There are already " + (numSnapshots + 1)
+          + " snapshot(s) and the " + type + " snapshot limit is " + limit;
       if (fsdir.isImageLoaded()) {
         // We have reached the maximum snapshot limit
         throw new SnapshotException(

```

**Review by szetszwo — COMMENTED** [review]
The change looks good.  Just a minor comment on the error message.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 12s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  33m 38s |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 16s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  compile  |   1m  8s |  trunk passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  checkstyle  |   0m 47s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 16s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  17m 44s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 50s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javadoc  |   1m 18s |  trunk passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +0 :ok: |  spotbugs  |   3m  7s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   3m  5s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 11s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javac  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  2s |  the patch passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  javac  |   1m  2s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 43s |  hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 89 unchanged - 0 fixed = 90 total (was 89)  |
| +1 :green_heart: |  mvnsite  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 41s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 45s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javadoc  |   1m 16s |  the patch passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  findbugs  |   3m 10s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 108m 47s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 36s |  The patch does not generate ASF License warnings.  |
|  |   | 199m 30s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.hdfs.TestFileChecksum |
|   | hadoop.hdfs.web.TestWebHdfsWithMultipleNameNodes |
|   | hadoop.hdfs.TestSnapshotCommands |
|   | hadoop.hdfs.server.namenode.ha.TestBootstrapAliasmap |
|   | hadoop.hdfs.server.namenode.ha.TestInitializeSharedEdits |
|   | hadoop.hdfs.server.namenode.snapshot.TestSnapshotManager |
|   | hadoop.hdfs.TestFileChecksumCompositeCrc |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2296 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux aa36c5473309 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 89428f142fe |
| Default Java | Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01 |
| checkstyle | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/2/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/2/testReport/ |
| Max. process+thread count | 2712 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/2/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 10s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  31m 51s |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 17s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  compile  |   1m 10s |  trunk passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  checkstyle  |   0m 50s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 15s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  17m 51s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 51s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javadoc  |   1m 25s |  trunk passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +0 :ok: |  spotbugs  |   3m  8s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   3m  6s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 11s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javac  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  4s |  the patch passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  javac  |   1m  4s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 41s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 10s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 34s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 45s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  javadoc  |   1m 16s |  the patch passed with JDK Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01  |
| +1 :green_heart: |  findbugs  |   3m 11s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 112m 34s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 37s |  The patch does not generate ASF License warnings.  |
|  |   | 201m 39s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.hdfs.TestGetFileChecksum |
|   | hadoop.hdfs.TestFileChecksumCompositeCrc |
|   | hadoop.hdfs.server.namenode.snapshot.TestSnapshotManager |
|   | hadoop.hdfs.TestBlockStoragePolicy |
|   | hadoop.hdfs.TestSnapshotCommands |
|   | hadoop.hdfs.TestFileChecksum |
|   | hadoop.hdfs.TestDFSOutputStream |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2296 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux b21b996bec02 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 2a6fdcedeb4 |
| Default Java | Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_265-8u265-b01-0ubuntu2~18.04-b01 |
| unit | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/3/testReport/ |
| Max. process+thread count | 2682 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2296/3/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java [hunk 8]
```diff
@@ -133,4 +137,58 @@ public void testValidateSnapshotIDWidth() throws Exception {
...
+            DFS_NAMENODE_SNAPSHOT_MAX_LIMIT, numSnapshots);
+    conf.setInt(DFSConfigKeys.DFS_NAMENODE_SNAPSHOT_FILESYSTEM_LIMIT,
+        numSnapshots * 2);
+    MiniDFSCluster cluster = new MiniDFSCluster.Builder(conf).
+        numDataNodes(0).build();
+    cluster.waitActive();
+    DistributedFileSystem hdfs = cluster.getFileSystem();
+    hdfs.mkdirs(snapshottableDir);
+    hdfs.allowSnapshot(snapshottableDir);
+    int i = 0;
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java:156]: Who uses this "i" later? in for loop def should be good.

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java [hunk 9]
```diff
@@ -133,4 +137,58 @@ public void testValidateSnapshotIDWidth() throws Exception {
...
+
+    // Any new snapshot creation should still fail
+    LambdaTestUtils.intercept(SnapshotException.class,
+        "snapshot limit",
+        () -> hdfs.createSnapshot(snapshottableDir, "s5"));
+    // now change max snapshot FS limit to 2 and restart namenode
+    cluster.getNameNode().getConf().setInt(DFSConfigKeys.
+        DFS_NAMENODE_SNAPSHOT_FILESYSTEM_LIMIT, 2);
+    cluster.restartNameNodes();
+    snapshotManager = cluster.getNamesystem().
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java:185]: 1 line?

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java [hunk 10]
```diff
@@ -133,4 +137,58 @@ public void testValidateSnapshotIDWidth() throws Exception {
...
+  public void testSnapshotLimitOnRestart() throws Exception {
+    final Configuration conf = new Configuration();
+    final Path snapshottableDir
+        = new Path("/" + getClass().getSimpleName());
+    int numSnapshots = 5;
+    conf.setInt(DFSConfigKeys.
+            DFS_NAMENODE_SNAPSHOT_MAX_LIMIT, numSnapshots);
+    conf.setInt(DFSConfigKeys.DFS_NAMENODE_SNAPSHOT_FILESYSTEM_LIMIT,
+        numSnapshots * 2);
+    MiniDFSCluster cluster = new MiniDFSCluster.Builder(conf).
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java:150]: should we clean this cluster?

**Review by goiri — APPROVED** [review]

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java [hunk 11]
```diff
@@ -133,4 +138,57 @@ public void testValidateSnapshotIDWidth() throws Exception {
...
+        DFS_NAMENODE_SNAPSHOT_FILESYSTEM_LIMIT, 2);
+    cluster.restartNameNodes();
+    snapshotManager = cluster.getNamesystem().
+        getSnapshotManager();
+    // make sure edits of all previous 5 create snapshots are replayed
+    Assert.assertEquals(numSnapshots, snapshotManager.getNumSnapshots());
+
+    // make sure namenode has the new snapshot limit configured as 2
+    Assert.assertEquals(2, snapshotManager.getMaxSnapshotLimit());
+    cluster.shutdown();
```
Comment (goiri) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestSnapshotManager.java:191]: We may need to do it in a finally to make sure we always clean it. Also check for null.

**Comment by bshashikant** [comment]
retest

**Comment by bshashikant** [comment]
Thanks @szetszwo and @goiri for the review.

**Pull request merged**